TriblerGUI/widgets/trustpage/js/data_processor.js

```diff
@@ -0,0 +1,326 @@
+ *
+ * @param {GraphResponseData} response - The response
+ * @param {Object} interim - The interim result (expects interim.nodes, interim.edges)
+ * @returns {{links: GraphLink[]}}
+ */
+function combineLinks(response, interim) {
+
+    var groupedEdges = groupBy(interim.edges, "source_pk");
+
+    // Make the combined links for all nodes
+    var combinedLinks = flatMap(interim.nodes, function (sourceNode) {
+        return (groupedEdges[sourceNode.public_key] || []).map(function (edge) {
+            var inverseEdge = find(groupedEdges[edge.target_pk], "target_pk", sourceNode.public_key);
+            var up = edge.amount,
+                down = inverseEdge ? inverseEdge.amount : 0,
+                empty = up + down === 0;
+
+            // Remove the inverse edge
+            if (inverseEdge) groupedEdges[edge.target_pk].splice(groupedEdges[edge.target_pk].indexOf(inverseEdge), 1);
+
+            return {
+                source_pk: edge.source_pk,
+                target_pk: edge.target_pk,
+                source: edge.source,
+                target: edge.target,
+                amount_up: up,
+                amount_down: down,
+                ratio: empty ? 0 : up / (up + down),
+                logRatio: empty ? 0 : Math.log(up + 1) / (Math.log(up + 1) + Math.log(down + 1)),
+            };
+        });
+    });
+
+    // Sort combined links ascending on up + down
+    combinedLinks.sort(function (linkOne, linkTwo) {
+        return linkOne.amount_up + linkOne.amount_down - linkTwo.amount_up - linkTwo.amount_down;
+    });
+
+    // Add reference to source and target object
+    return {
+        links: combinedLinks
+    };
+}
+
+/**
+ * Make a map from local_keys (integers 0 to n-1) to public keys (where n is the number of nodes).
+ * (Manipulates the interim)
+ *
+ * @param {GraphResponseData} response - The response
+ * @param {Object} interim - The interim result (expects interim.nodes)
+ * @returns {{local_keys: String[]}}
+ */
+function makeLocalKeyMap(response, interim) {
+
+    var localKeys = interim.nodes.map(function (node, i) {
+        node.local_key = i;
+        return node.public_key;
+    });
+
+    return {
+        local_keys: localKeys
+    };
+}
+
+/**
+ * Get the focus node object and its public key.
+ *
+ * @param {GraphResponseData} response - The response
+ * @param {Object} interim - The interim result (expects interim.nodes)
+ * @returns {{focus_node: GraphNode, focus_pk: String}}
+ */
+function focusNodePublicKey(response, interim) {
+
+    return {
+        focus_node: find(interim.nodes, "public_key", response.focus_node),
+        focus_pk: response.focus_node
+    };
+}
+
+/**
+ * Add the neighbors to each node. (Manipulates the interim)
+ *
+ * @param {GraphResponseData} response - The response
+ * @param {Object} interim - The interim result (expects interim.links, interim.nodes)
+ * @returns {{}} - An empty object, since this manipulates the interim.
```

Comment:
I don't really like it either. But I don't see an elegant way of doing it purely functional.